File: Chronir-Android/core/designsystem/src/main/java/com/chronir/designsystem/atoms/ChronirText.kt
package com.chronir.designsystem.atoms

import android.content.res.Configuration.UI_MODE_NIGHT_YES
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.chronir.designsystem.theme.ChronirTheme
import com.chronir.designsystem.tokens.TypographyTokens

enum class ChronirTextStyle {
    // Spec-aligned (design-system.md Section 3.3)
    DisplayAlarm,
    HeadlineTime,
    HeadlineTitle,
    BodyPrimary,
    BodySecondary,
    CaptionCountdown,
    CaptionBadge,

    // Full Material-style scale
    DisplayLarge,
    DisplayMedium,
    DisplaySmall,
    HeadlineLarge,
    HeadlineMedium,
    HeadlineSmall,
    TitleLarge,
    TitleMedium,
    TitleSmall,
    BodyLarge,
    BodyMedium,
    BodySmall,
    LabelLarge,
    LabelMedium,
    LabelSmall,
    Caption;

    // Static fallback — only used outside a composable scope
    fun toTextStyle(): TextStyle = when (this) {
        DisplayAlarm -> TypographyTokens.DisplayAlarm
        HeadlineTime -> TypographyTokens.HeadlineTime
        HeadlineTitle -> TypographyTokens.HeadlineTitle
        BodyPrimary -> TypographyTokens.BodyPrimary
        BodySecondary -> TypographyTokens.BodySecondary
        CaptionCountdown -> TypographyTokens.CaptionCountdown
        CaptionBadge -> TypographyTokens.CaptionBadge
        DisplayLarge -> TypographyTokens.DisplayLarge
        DisplayMedium -> TypographyTokens.DisplayMedium
        DisplaySmall -> TypographyTokens.DisplaySmall
        HeadlineLarge -> TypographyTokens.HeadlineLarge
        HeadlineMedium -> TypographyTokens.HeadlineMedium
        HeadlineSmall -> TypographyTokens.HeadlineSmall
        TitleLarge -> TypographyTokens.TitleLarge
        TitleMedium -> TypographyTokens.TitleMedium
        TitleSmall -> TypographyTokens.TitleSmall
        BodyLarge -> TypographyTokens.BodyLarge
        BodyMedium -> TypographyTokens.BodyMedium
        BodySmall -> TypographyTokens.BodySmall
        LabelLarge -> TypographyTokens.LabelLarge
        LabelMedium -> TypographyTokens.LabelMedium
        LabelSmall -> TypographyTokens.LabelSmall
        Caption -> TypographyTokens.Caption
    }
}

/**
 * Resolves the ChronirTextStyle from MaterialTheme.typography so text scaling is applied.
 * DisplayAlarm and Caption fall back to static tokens since they have no M3 equivalent.
 */
@Composable
private fun ChronirTextStyle.resolve(): TextStyle {
    val typography = MaterialTheme.typography
    return when (this) {
        // Spec-aligned aliases resolve through their M3 mapping
        ChronirTextStyle.DisplayAlarm -> TypographyTokens.DisplayAlarm // 120sp — no M3 equivalent, stays unscaled
        ChronirTextStyle.HeadlineTime -> typography.headlineLarge
        ChronirTextStyle.HeadlineTitle -> typography.headlineSmall
        ChronirTextStyle.BodyPrimary -> typography.bodyLarge
        ChronirTextStyle.BodySecondary -> typography.bodyMedium
        ChronirTextStyle.CaptionCountdown -> typography.labelLarge
        ChronirTextStyle.CaptionBadge -> typography.labelMedium
        // Full M3 scale
        ChronirTextStyle.DisplayLarge -> typography.displayLarge
        ChronirTextStyle.DisplayMedium -> typography.displayMedium
        ChronirTextStyle.DisplaySmall -> typography.displayMedium // displaySmall maps to our DisplaySmall via AppTypography
        ChronirTextStyle.HeadlineLarge -> typography.headlineLarge
        ChronirTextStyle.HeadlineMedium -> typography.headlineMedium
        ChronirTextStyle.HeadlineSmall -> typography.headlineSmall
        ChronirTextStyle.TitleLarge -> typography.titleLarge
        ChronirTextStyle.TitleMedium -> typography.titleMedium
        ChronirTextStyle.TitleSmall -> typography.titleSmall
        ChronirTextStyle.BodyLarge -> typography.bodyLarge
        ChronirTextStyle.BodyMedium -> typography.bodyMedium
        ChronirTextStyle.BodySmall -> typography.bodySmall
        ChronirTextStyle.LabelLarge -> typography.labelLarge
        ChronirTextStyle.LabelMedium -> typography.labelMedium
        ChronirTextStyle.LabelSmall -> typography.labelSmall
        ChronirTextStyle.Caption -> TypographyTokens.Caption // 10sp — no M3 equivalent
    }
}

@Composable
fun ChronirText(
    text: String,
    modifier: Modifier = Modifier,
    style: ChronirTextStyle = ChronirTextStyle.BodyPrimary,
    color: Color = Color.Unspecified,
    maxLines: Int = Int.MAX_VALUE,
    overflow: TextOverflow = TextOverflow.Clip,
    textAlign: TextAlign? = null
) {
    Text(
        text = text,
        modifier = modifier,
        style = style.resolve(),
        color = color,
        maxLines = maxLines,
        overflow = overflow,
        textAlign = textAlign
    )
}

@Preview(name = "Spec-Aligned — Light", showBackground = true)
@Preview(name = "Spec-Aligned — Dark", showBackground = true, uiMode = UI_MODE_NIGHT_YES)
@Composable
private fun ChronirTextSpecPreview() {
    ChronirTheme(dynamicColor = false) {
        Column(modifier = Modifier.padding(16.dp)) {
            ChronirText(text = "12:00", style = ChronirTextStyle.DisplayAlarm)
            ChronirText(text = "3:45 PM", style = ChronirTextStyle.HeadlineTime)
            ChronirText(text = "Screen Title", style = ChronirTextStyle.HeadlineTitle)
            ChronirText(text = "Primary body text", style = ChronirTextStyle.BodyPrimary)
            ChronirText(
                text = "Secondary metadata",
                style = ChronirTextStyle.BodySecondary,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            ChronirText(text = "Alarm in 6h 32m", style = ChronirTextStyle.CaptionCountdown)
            ChronirText(
                text = "Weekly",
                style = ChronirTextStyle.CaptionBadge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Preview(name = "Large Font", showBackground = true, fontScale = 2f)
@Composable
private fun ChronirTextLargeFontPreview() {
    ChronirTheme(dynamicColor = false) {
        Column(modifier = Modifier.padding(16.dp)) {
            ChronirText(text = "Screen Title", style = ChronirTextStyle.HeadlineTitle)
            ChronirText(text = "Primary body text", style = ChronirTextStyle.BodyPrimary)
            ChronirText(text = "Caption badge", style = ChronirTextStyle.CaptionBadge)
        }
    }
}

File: chronir/chronir/DesignSystem/Atoms/ChronirText.swift
import SwiftUI

// MARK: - Environment Key

private struct TextSizeScaleKey: EnvironmentKey {
    static let defaultValue: CGFloat = 1.0
}

extension EnvironmentValues {
    var textSizeScale: CGFloat {
        get { self[TextSizeScaleKey.self] }
        set { self[TextSizeScaleKey.self] = newValue }
    }
}

// MARK: - Font Attributes

struct FontAttributes {
    let size: CGFloat
    let weight: Font.Weight
    let design: Font.Design
}

// MARK: - Text Style

enum ChronirTextStyle {
    // Spec-aligned (design-system.md Section 3.3)
    case displayAlarm
    case headlineTime
    case headlineTitle
    case bodyPrimary
    case bodySecondary
    case captionCountdown
    case captionBadge

    // Full Material-style scale
    case displayLarge
    case displayMedium
    case displaySmall
    case headlineLarge
    case headlineMedium
    case headlineSmall
    case titleLarge
    case titleMedium
    case titleSmall
    case bodyLarge
    case bodyMedium
    case bodySmall
    case labelLarge
    case labelMedium
    case labelSmall
    case caption

    /// Base size, weight, and design for each style (scale = 1.0).
    var baseAttributes: FontAttributes {
        switch self {
        // Spec-aligned aliases resolve to their underlying token values
        case .displayAlarm, .displayLarge: FontAttributes(size: 120, weight: .bold, design: .rounded)
        case .displayMedium: FontAttributes(size: 57, weight: .bold, design: .rounded)
        case .displaySmall: FontAttributes(size: 36, weight: .bold, design: .rounded)
        case .headlineLarge: FontAttributes(size: 32, weight: .semibold, design: .default)
        case .headlineMedium: FontAttributes(size: 28, weight: .semibold, design: .default)
        case .headlineSmall, .headlineTime, .headlineTitle:
            FontAttributes(size: 24, weight: .semibold, design: .default)
        case .titleLarge: FontAttributes(size: 22, weight: .medium, design: .default)
        case .titleMedium: FontAttributes(size: 18, weight: .medium, design: .default)
        case .titleSmall: FontAttributes(size: 16, weight: .medium, design: .default)
        case .bodyLarge, .bodyPrimary: FontAttributes(size: 16, weight: .regular, design: .default)
        case .bodyMedium, .bodySecondary: FontAttributes(size: 14, weight: .regular, design: .default)
        case .bodySmall: FontAttributes(size: 12, weight: .regular, design: .default)
        case .labelLarge, .captionCountdown: FontAttributes(size: 14, weight: .medium, design: .default)
        case .labelMedium, .captionBadge: FontAttributes(size: 12, weight: .medium, design: .default)
        case .labelSmall: FontAttributes(size: 11, weight: .medium, design: .default)
        case .caption: FontAttributes(size: 10, weight: .regular, design: .default)
        }
    }

    /// Unscaled font (scale = 1.0). Used by widgets and backward-compat paths.
    var font: Font {
        let attr = baseAttributes
        return Font.system(size: attr.size, weight: attr.weight, design: attr.design)
    }

    /// Scaled font using the given scale factor.
    func font(scale: CGFloat) -> Font {
        let attr = baseAttributes
        return TypographyTokens.scaled(size: attr.size, weight: attr.weight, design: attr.design, scale: scale)
    }
}

// MARK: - ChronirText View

struct ChronirText: View {
    let text: String
    var style: ChronirTextStyle
    var color: Color
    var maxLines: Int?
    var alignment: TextAlignment

    @Environment(\.textSizeScale) private var textSizeScale

    init(
        _ text: String,
        style: ChronirTextStyle = .bodyPrimary,
        color: Color = ColorTokens.textPrimary,
        maxLines: Int? = nil,
        alignment: TextAlignment = .leading
    ) {
        self.text = text
        self.style = style
        self.color = color
        self.maxLines = maxLines
        self.alignment = alignment
    }

    var body: some View {
        Text(text)
            .font(style.font(scale: textSizeScale))
            .foregroundStyle(color)
            .lineLimit(maxLines)
            .multilineTextAlignment(alignment)
    }
}

// MARK: - View Modifier for Direct .font() Call Sites

struct ChronirFontModifier: ViewModifier {
    @Environment(\.textSizeScale) private var scale
    let style: ChronirTextStyle

    func body(content: Content) -> some View {
        content.font(style.font(scale: scale))
    }
}

extension View {
    func chronirFont(_ style: ChronirTextStyle) -> some View {
        modifier(ChronirFontModifier(style: style))
    }
}

// MARK: - Previews

#Preview("Spec-Aligned Variants") {
    VStack(alignment: .leading, spacing: SpacingTokens.md) {
        ChronirText("12:00", style: .displayAlarm)
        ChronirText("3:45 PM", style: .headlineTime)
        ChronirText("Screen Title", style: .headlineTitle)
        ChronirText("Primary body text", style: .bodyPrimary)
        ChronirText("Secondary metadata", style: .bodySecondary, color: ColorTokens.textSecondary)
        ChronirText("Alarm in 6h 32m", style: .captionCountdown)
        ChronirText("Weekly", style: .captionBadge, color: ColorTokens.textSecondary)
    }
    .padding()
    .background(ColorTokens.backgroundPrimary)
}

#Preview("Full Type Scale") {
    ScrollView {
        VStack(alignment: .leading, spacing: SpacingTokens.sm) {
            ChronirText("Display Large", style: .displayLarge)
            ChronirText("Display Medium", style: .displayMedium)
            ChronirText("Display Small", style: .displaySmall)
            ChronirText("Headline Large", style: .headlineLarge)
            ChronirText("Headline Medium", style: .headlineMedium)
            ChronirText("Headline Small", style: .headlineSmall)
            ChronirText("Title Large", style: .titleLarge)
            ChronirText("Title Medium", style: .titleMedium)
            ChronirText("Title Small", style: .titleSmall)
            ChronirText("Body Large", style: .bodyLarge)
            ChronirText("Body Medium", style: .bodyMedium)
            ChronirText("Body Small", style: .bodySmall)
            ChronirText("Label Large", style: .labelLarge)
            ChronirText("Label Medium", style: .labelMedium)
            ChronirText("Label Small", style: .labelSmall)
            ChronirText("Caption", style: .caption)
        }
        .padding()
    }
    .background(ColorTokens.backgroundPrimary)
}
